File: backend/app/schemas/cdp.py
# =============================================================================
# Stratum AI - CDP Pydantic Schemas
# =============================================================================
"""
Pydantic schemas for CDP API request/response validation.

Schemas:
- Event ingestion (single + batch)
- Profile responses
- Source management
- Consent handling
"""

from datetime import datetime
from decimal import Decimal
from typing import Optional, List, Dict, Any
from uuid import UUID

from pydantic import BaseModel, Field, field_validator, ConfigDict


# =============================================================================
# Base Configuration
# =============================================================================

class CDPBaseSchema(BaseModel):
    """Base schema with common configuration."""
    model_config = ConfigDict(
        from_attributes=True,
        str_strip_whitespace=True,
        validate_assignment=True,
    )


# =============================================================================
# Identifier Schemas
# =============================================================================

class IdentifierInput(BaseModel):
    """Single identifier in an event."""
    type: str = Field(..., description="Identifier type: email, phone, device_id, anonymous_id, external_id")
    value: str = Field(..., min_length=1, max_length=512, description="Identifier value")

    @field_validator("type")
    @classmethod
    def validate_type(cls, v: str) -> str:
        allowed = {"email", "phone", "device_id", "anonymous_id", "external_id"}
        if v.lower() not in allowed:
            raise ValueError(f"Invalid identifier type. Allowed: {allowed}")
        return v.lower()


class IdentifierResponse(CDPBaseSchema):
    """Identifier in API responses."""
    id: UUID
    identifier_type: str
    identifier_value: Optional[str] = None
    identifier_hash: str
    is_primary: bool
    confidence_score: Decimal
    verified_at: Optional[datetime] = None
    first_seen_at: datetime
    last_seen_at: datetime


# =============================================================================
# Event Schemas
# =============================================================================

class EventContext(BaseModel):
    """Context data for an event (device, geo, campaign info)."""
    user_agent: Optional[str] = Field(None, max_length=1024)
    ip: Optional[str] = Field(None, max_length=45)
    locale: Optional[str] = Field(None, max_length=20)
    timezone: Optional[str] = Field(None, max_length=50)
    screen: Optional[Dict[str, int]] = None
    campaign: Optional[Dict[str, str]] = None


class EventConsent(BaseModel):
    """Consent flags sent with event."""
    analytics: Optional[bool] = None
    ads: Optional[bool] = None
    email: Optional[bool] = None
    sms: Optional[bool] = None


class EventInput(BaseModel):
    """Single event for ingestion."""
    event_name: str = Field(..., min_length=1, max_length=255, description="Event name (e.g., PageView, Purchase)")
    event_time: datetime = Field(..., description="When the event occurred (ISO8601)")
    idempotency_key: Optional[str] = Field(None, max_length=128, description="Unique key for deduplication")

    identifiers: List[IdentifierInput] = Field(..., min_length=1, description="At least one identifier required")

    properties: Optional[Dict[str, Any]] = Field(default_factory=dict, description="Event-specific properties")
    context: Optional[EventContext] = Field(default=None, description="Device, geo, campaign context")
    consent: Optional[EventConsent] = Field(default=None, description="Consent flags")

    @field_validator("event_name")
    @classmethod
    def validate_event_name(cls, v: str) -> str:
        # Allow alphanumeric, spaces, underscores, dashes
        import re
        if not re.match(r"^[\w\s\-]+$", v):
            raise ValueError("Event name can only contain letters, numbers, spaces, underscores, and dashes")
        return v


class EventBatchInput(BaseModel):
    """Batch of events for ingestion."""
    events: List[EventInput] = Field(..., min_length=1, max_length=1000, description="Events to ingest (max 1000)")


class EventIngestResult(BaseModel):
    """Result for a single event ingestion."""
    event_id: Optional[UUID] = None
    status: str  # "accepted", "rejected", "duplicate"
    profile_id: Optional[UUID] = None
    error: Optional[str] = None


class EventBatchResponse(BaseModel):
    """Response for batch event ingestion."""
    accepted: int
    rejected: int
    duplicates: int
    results: List[EventIngestResult]


class EventResponse(CDPBaseSchema):
    """Event in API responses."""
    id: UUID
    event_name: str
    event_time: datetime
    received_at: datetime
    properties: Dict[str, Any]
    context: Dict[str, Any]
    emq_score: Optional[Decimal] = None
    processed: bool


# =============================================================================
# Profile Schemas
# =============================================================================

class ProfileResponse(CDPBaseSchema):
    """Full profile response with identifiers and metadata."""
    id: UUID
    tenant_id: int
    external_id: Optional[str] = None

    first_seen_at: datetime
    last_seen_at: datetime

    profile_data: Dict[str, Any]
    computed_traits: Dict[str, Any]

    lifecycle_stage: str

    total_events: int
    total_sessions: int
    total_purchases: int
    total_revenue: Decimal

    identifiers: List[IdentifierResponse] = []

    created_at: datetime
    updated_at: datetime


class ProfileListResponse(BaseModel):
    """Paginated list of profiles."""
    profiles: List[ProfileResponse]
    total: int
    page: int
    page_size: int


class ProfileLookupParams(BaseModel):
    """Parameters for profile lookup by identifier."""
    identifier_type: str = Field(..., description="Type of identifier (email, phone, etc.)")
    identifier_value: str = Field(..., description="Value to look up")


# =============================================================================
# Source Schemas
# =============================================================================

class SourceCreate(BaseModel):
    """Create a new data source."""
    name: str = Field(..., min_length=1, max_length=255, description="Human-readable source name")
    source_type: str = Field(..., description="Source type: website, server, sgtm, import, crm")
    config: Optional[Dict[str, Any]] = Field(default_factory=dict, description="Source-specific configuration")

    @field_validator("source_type")
    @classmethod
    def validate_source_type(cls, v: str) -> str:
        allowed = {"website", "server", "sgtm", "import", "crm"}
        if v.lower() not in allowed:
            raise ValueError(f"Invalid source type. Allowed: {allowed}")
        return v.lower()


class SourceResponse(CDPBaseSchema):
    """Data source in API responses."""
    id: UUID
    name: str
    source_type: str
    source_key: str
    config: Dict[str, Any]
    is_active: bool
    event_count: int
    last_event_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime


class SourceListResponse(BaseModel):
    """List of data sources."""
    sources: List[SourceResponse]
    total: int


# =============================================================================
# Consent Schemas
# =============================================================================

class ConsentUpdate(BaseModel):
    """Update consent for a profile."""
    consent_type: str = Field(..., description="Consent type: analytics, ads, email, sms, all")
    granted: bool = Field(..., description="Whether consent is granted")
    source: Optional[str] = Field(None, description="Where consent was collected")
    consent_text: Optional[str] = Field(None, description="Legal text shown to user")
    consent_version: Optional[str] = Field(None, description="Version of consent form")

    @field_validator("consent_type")
    @classmethod
    def validate_consent_type(cls, v: str) -> str:
        allowed = {"analytics", "ads", "email", "sms", "all"}
        if v.lower() not in allowed:
            raise ValueError(f"Invalid consent type. Allowed: {allowed}")
        return v.lower()


class ConsentResponse(CDPBaseSchema):
    """Consent record in API responses."""
    id: UUID
    consent_type: str
    granted: bool
    granted_at: Optional[datetime] = None
    revoked_at: Optional[datetime] = None
    source: Optional[str] = None
    consent_version: Optional[str] = None
    created_at: datetime
    updated_at: datetime


# =============================================================================
# EMQ (Event Match Quality) Schemas
# =============================================================================

class EMQScoreResponse(BaseModel):
    """EMQ score breakdown for an event or batch."""
    overall_score: Decimal = Field(..., description="Overall EMQ score (0-100)")
    identifier_quality: Decimal = Field(..., description="Score for identifier quality (0-40)")
    data_completeness: Decimal = Field(..., description="Score for data completeness (0-25)")
    timeliness: Decimal = Field(..., description="Score for event timeliness (0-20)")
    context_richness: Decimal = Field(..., description="Score for context data (0-15)")


# =============================================================================
# API Response Wrappers
# =============================================================================

class CDPAPIResponse(BaseModel):
    """Standard API response wrapper."""
    success: bool = True
    data: Optional[Any] = None
    message: Optional[str] = None
    errors: Optional[List[Dict[str, Any]]] = None


# =============================================================================
# Webhook Schemas
# =============================================================================

class WebhookCreate(BaseModel):
    """Create a new webhook destination."""
    name: str = Field(..., min_length=1, max_length=255, description="Human-readable webhook name")
    url: str = Field(..., min_length=1, max_length=2048, description="Webhook destination URL (HTTPS)")
    event_types: List[str] = Field(
        ...,
        min_length=1,
        description="Event types to trigger on: event.received, profile.created, profile.updated, profile.merged, consent.updated, all"
    )
    max_retries: int = Field(3, ge=0, le=10, description="Max retry attempts on failure")
    timeout_seconds: int = Field(30, ge=5, le=120, description="Request timeout in seconds")

    @field_validator("url")
    @classmethod
    def validate_url(cls, v: str) -> str:
        if not v.startswith(("https://", "http://localhost")):
            raise ValueError("Webhook URL must use HTTPS (http://localhost allowed for testing)")
        return v

    @field_validator("event_types")
    @classmethod
    def validate_event_types(cls, v: List[str]) -> List[str]:
        allowed = {"event.received", "profile.created", "profile.updated", "profile.merged", "consent.updated", "all"}
        for event_type in v:
            if event_type not in allowed:
                raise ValueError(f"Invalid event type '{event_type}'. Allowed: {allowed}")
        return v


class WebhookUpdate(BaseModel):
    """Update an existing webhook."""
    name: Optional[str] = Field(None, min_length=1, max_length=255)
    url: Optional[str] = Field(None, min_length=1, max_length=2048)
    event_types: Optional[List[str]] = None
    is_active: Optional[bool] = None
    max_retries: Optional[int] = Field(None, ge=0, le=10)
    timeout_seconds: Optional[int] = Field(None, ge=5, le=120)

    @field_validator("url")
    @classmethod
    def validate_url(cls, v: Optional[str]) -> Optional[str]:
        if v and not v.startswith(("https://", "http://localhost")):
            raise ValueError("Webhook URL must use HTTPS (http://localhost allowed for testing)")
        return v

    @field_validator("event_types")
    @classmethod
    def validate_event_types(cls, v: Optional[List[str]]) -> Optional[List[str]]:
        if v is None:
            return v
        allowed = {"event.received", "profile.created", "profile.updated", "profile.merged", "consent.updated", "all"}
        for event_type in v:
            if event_type not in allowed:
                raise ValueError(f"Invalid event type '{event_type}'. Allowed: {allowed}")
        return v


class WebhookResponse(CDPBaseSchema):
    """Webhook in API responses."""
    id: UUID
    name: str
    url: str
    event_types: List[str]
    secret_key: Optional[str] = None  # Only returned on create
    is_active: bool
    last_triggered_at: Optional[datetime] = None
    last_success_at: Optional[datetime] = None
    last_failure_at: Optional[datetime] = None
    failure_count: int
    max_retries: int
    timeout_seconds: int
    created_at: datetime
    updated_at: datetime


class WebhookListResponse(BaseModel):
    """List of webhooks."""
    webhooks: List[WebhookResponse]
    total: int


class WebhookTestResult(BaseModel):
    """Result of webhook test request."""
    success: bool
    status_code: Optional[int] = None
    response_time_ms: Optional[float] = None
    error: Optional[str] = None

File: frontend/src/api/cdp.ts
/**
 * Stratum AI - CDP (Customer Data Platform) API Client
 *
 * API client with TypeScript types and React Query hooks for CDP endpoints.
 * Handles event ingestion, profile lookups, and source management.
 */

import { useQuery, useMutation, useQueryClient } from '@tanstack/react-query'
import { apiClient, ApiResponse } from './client'

// =============================================================================
// TypeScript Types
// =============================================================================

// Identifier Types
export type IdentifierType = 'email' | 'phone' | 'device_id' | 'anonymous_id' | 'external_id'

export interface IdentifierInput {
  type: IdentifierType
  value: string
}

export interface Identifier {
  id: string
  identifier_type: IdentifierType
  identifier_value?: string
  identifier_hash: string
  is_primary: boolean
  confidence_score: number
  verified_at?: string
  first_seen_at: string
  last_seen_at: string
}

// Event Types
export interface EventContext {
  user_agent?: string
  ip?: string
  locale?: string
  timezone?: string
  screen?: { width: number; height: number }
  campaign?: Record<string, string>
}

export interface EventConsent {
  analytics?: boolean
  ads?: boolean
  email?: boolean
  sms?: boolean
}

export interface EventInput {
  event_name: string
  event_time: string
  idempotency_key?: string
  identifiers: IdentifierInput[]
  properties?: Record<string, unknown>
  context?: EventContext
  consent?: EventConsent
}

export interface EventBatchInput {
  events: EventInput[]
}

export interface EventIngestResult {
  event_id?: string
  status: 'accepted' | 'rejected' | 'duplicate'
  profile_id?: string
  error?: string
}

export interface EventBatchResponse {
  accepted: number
  rejected: number
  duplicates: number
  results: EventIngestResult[]
}

export interface CDPEvent {
  id: string
  event_name: string
  event_time: string
  received_at: string
  properties: Record<string, unknown>
  context: Record<string, unknown>
  emq_score?: number
  processed: boolean
}

// Profile Types
export type LifecycleStage = 'anonymous' | 'known' | 'customer' | 'churned'

export interface CDPProfile {
  id: string
  tenant_id: number
  external_id?: string
  first_seen_at: string
  last_seen_at: string
  profile_data: Record<string, unknown>
  computed_traits: Record<string, unknown>
  lifecycle_stage: LifecycleStage
  total_events: number
  total_sessions: number
  total_purchases: number
  total_revenue: number
  identifiers: Identifier[]
  created_at: string
  updated_at: string
}

export interface ProfileListResponse {
  profiles: CDPProfile[]
  total: number
  page: number
  page_size: number
}

// Source Types
export type SourceType = 'website' | 'server' | 'sgtm' | 'import' | 'crm'

export interface SourceCreate {
  name: string
  source_type: SourceType
  config?: Record<string, unknown>
}

export interface CDPSource {
  id: string
  name: string
  source_type: SourceType
  source_key: string
  config: Record<string, unknown>
  is_active: boolean
  event_count: number
  last_event_at?: string
  created_at: string
  updated_at: string
}

export interface SourceListResponse {
  sources: CDPSource[]
  total: number
}

// Consent Types
export type ConsentType = 'analytics' | 'ads' | 'email' | 'sms' | 'all'

export interface ConsentUpdate {
  consent_type: ConsentType
  granted: boolean
  source?: string
  consent_text?: string
  consent_version?: string
}

export interface CDPConsent {
  id: string
  consent_type: ConsentType
  granted: boolean
  granted_at?: string
  revoked_at?: string
  source?: string
  consent_version?: string
  created_at: string
  updated_at: string
}

// EMQ Score Types
export interface EMQScore {
  overall_score: number
  identifier_quality: number
  data_completeness: number
  timeliness: number
  context_richness: number
}

// Health Check Response
export interface CDPHealthResponse {
  status: 'healthy' | 'degraded' | 'unhealthy'
  module: string
  version: string
}

// Webhook Types
export type WebhookEventType =
  | 'event.received'
  | 'profile.created'
  | 'profile.updated'
  | 'profile.merged'
  | 'consent.updated'
  | 'all'

export interface WebhookCreate {
  name: string
  url: string
  event_types: WebhookEventType[]
  max_retries?: number
  timeout_seconds?: number
}

export interface WebhookUpdate {
  name?: string
  url?: string
  event_types?: WebhookEventType[]
  is_active?: boolean
  max_retries?: number
  timeout_seconds?: number
}

export interface CDPWebhook {
  id: string
  name: string
  url: string
  event_types: WebhookEventType[]
  secret_key?: string // Only returned on create/rotate
  is_active: boolean
  last_triggered_at?: string
  last_success_at?: string
  last_failure_at?: string
  failure_count: number
  max_retries: number
  timeout_seconds: number
  created_at: string
  updated_at: string
}

export interface WebhookListResponse {
  webhooks: CDPWebhook[]
  total: number
}

export interface WebhookTestResult {
  success: boolean
  status_code?: number
  response_time_ms?: number
  error?: string
}

// Anomaly Detection Types
export type AnomalySeverity = 'low' | 'medium' | 'high' | 'critical'
export type AnomalyDirection = 'high' | 'low'

export interface EventAnomaly {
  source_id?: string
  source_name: string
  metric: string
  zscore: number
  severity: AnomalySeverity
  current_value: number
  baseline_mean: number
  baseline_std: number
  direction: AnomalyDirection
  pct_change: number
}

export interface AnomalyDetectionResponse {
  anomalies: EventAnomaly[]
  anomaly_count: number
  has_critical: boolean
  has_high: boolean
  analysis_period_days: number
  zscore_threshold: number
  total_sources_analyzed: number
}

export interface AnomalyDetectionParams {
  window_days?: number
  zscore_threshold?: number
}

export type HealthStatus = 'healthy' | 'fair' | 'degraded' | 'critical' | 'unknown'
export type VolumeTrend = 'increasing' | 'stable' | 'decreasing'

export interface AnomalySummaryResponse {
  health_status: HealthStatus
  events_today: number
  events_7d: number
  events_prev_7d: number
  wow_change_pct: number
  volume_trend: VolumeTrend
  avg_emq_score?: number
  as_of: string
}

// Export Types
export type ExportFormat = 'json' | 'csv'

export interface ExportProfilesParams {
  format?: ExportFormat
  limit?: number
  offset?: number
}

export interface ExportEventsParams {
  format?: ExportFormat
  limit?: number
  offset?: number
  start_date?: string
  end_date?: string
  event_name?: string
}

export interface ExportResponse<T> {
  format: string
  count: number
  offset: number
  limit: number
  data: T[]
  filters?: Record<string, string | null>
}

// =============================================================================
// API Functions
// =============================================================================

export const cdpApi = {
  // Event Endpoints
  /**
   * Ingest events into CDP
   * @param events Array of events to ingest
   * @param sourceKey Optional source API key for authentication
   */
  ingestEvents: async (
    events: EventInput[],
    sourceKey?: string
  ): Promise<EventBatchResponse> => {
    const params = sourceKey ? { source_key: sourceKey } : {}
    const response = await apiClient.post<ApiResponse<EventBatchResponse>>(
      '/cdp/events',
      { events },
      { params }
    )
    return response.data.data
  },

  /**
   * Ingest a single event
   * @param event Event to ingest
   * @param sourceKey Optional source API key
   */
  ingestEvent: async (
    event: EventInput,
    sourceKey?: string
  ): Promise<EventIngestResult> => {
    const response = await cdpApi.ingestEvents([event], sourceKey)
    return response.results[0]
  },

  // Profile Endpoints
  /**
   * Get a profile by ID
   * @param profileId UUID of the profile
   */
  getProfile: async (profileId: string): Promise<CDPProfile> => {
    const response = await apiClient.get<ApiResponse<CDPProfile>>(
      `/cdp/profiles/${profileId}`
    )
    return response.data.data
  },

  /**
   * Lookup a profile by identifier
   * @param identifierType Type of identifier (email, phone, etc.)
   * @param identifierValue Value to look up
   */
  lookupProfile: async (
    identifierType: IdentifierType,
    identifierValue: string
  ): Promise<CDPProfile> => {
    const response = await apiClient.get<ApiResponse<CDPProfile>>(
      '/cdp/profiles',
      {
        params: {
          identifier_type: identifierType,
          identifier_value: identifierValue,
        },
      }
    )
    return response.data.data
  },

  // Source Endpoints
  /**
   * List all data sources
   */
  listSources: async (): Promise<SourceListResponse> => {
    const response = await apiClient.get<ApiResponse<SourceListResponse>>(
      '/cdp/sources'
    )
    return response.data.data
  },

  /**
   * Create a new data source
   * @param source Source configuration
   */
  createSource: async (source: SourceCreate): Promise<CDPSource> => {
    const response = await apiClient.post<ApiResponse<CDPSource>>(
      '/cdp/sources',
      source
    )
    return response.data.data
  },

  // Health Check
  /**
   * Check CDP module health
   */
  health: async (): Promise<CDPHealthResponse> => {
    const response = await apiClient.get<CDPHealthResponse>('/cdp/health')
    return response.data
  },

  // Webhook Endpoints
  /**
   * List all webhooks
   */
  listWebhooks: async (): Promise<WebhookListResponse> => {
    const response = await apiClient.get<ApiResponse<WebhookListResponse>>(
      '/cdp/webhooks'
    )
    return response.data.data
  },

  /**
   * Create a new webhook
   * @param webhook Webhook configuration
   */
  createWebhook: async (webhook: WebhookCreate): Promise<CDPWebhook> => {
    const response = await apiClient.post<ApiResponse<CDPWebhook>>(
      '/cdp/webhooks',
      webhook
    )
    return response.data.data
  },

  /**
   * Get a webhook by ID
   * @param webhookId UUID of the webhook
   */
  getWebhook: async (webhookId: string): Promise<CDPWebhook> => {
    const response = await apiClient.get<ApiResponse<CDPWebhook>>(
      `/cdp/webhooks/${webhookId}`
    )
    return response.data.data
  },

  /**
   * Update a webhook
   * @param webhookId UUID of the webhook
   * @param update Fields to update
   */
  updateWebhook: async (
    webhookId: string,
    update: WebhookUpdate
  ): Promise<CDPWebhook> => {
    const response = await apiClient.patch<ApiResponse<CDPWebhook>>(
      `/cdp/webhooks/${webhookId}`,
      update
    )
    return response.data.data
  },

  /**
   * Delete a webhook
   * @param webhookId UUID of the webhook
   */
  deleteWebhook: async (webhookId: string): Promise<void> => {
    await apiClient.delete(`/cdp/webhooks/${webhookId}`)
  },

  /**
   * Test a webhook
   * @param webhookId UUID of the webhook
   */
  testWebhook: async (webhookId: string): Promise<WebhookTestResult> => {
    const response = await apiClient.post<ApiResponse<WebhookTestResult>>(
      `/cdp/webhooks/${webhookId}/test`
    )
    return response.data.data
  },

  /**
   * Rotate webhook secret
   * @param webhookId UUID of the webhook
   */
  rotateWebhookSecret: async (webhookId: string): Promise<CDPWebhook> => {
    const response = await apiClient.post<ApiResponse<CDPWebhook>>(
      `/cdp/webhooks/${webhookId}/rotate-secret`
    )
    return response.data.data
  },

  // Anomaly Detection Endpoints
  /**
   * Detect event volume anomalies
   * @param params Optional parameters for analysis
   */
  detectEventAnomalies: async (
    params?: AnomalyDetectionParams
  ): Promise<AnomalyDetectionResponse> => {
    const response = await apiClient.get<AnomalyDetectionResponse>(
      '/cdp/anomalies/events',
      { params }
    )
    return response.data
  },

  /**
   * Get anomaly detection summary
   */
  getAnomalySummary: async (): Promise<AnomalySummaryResponse> => {
    const response = await apiClient.get<AnomalySummaryResponse>(
      '/cdp/anomalies/summary'
    )
    return response.data
  },

  // Export Endpoints
  /**
   * Export profiles in JSON or CSV format
   * @param params Export parameters (format, limit, offset)
   */
  exportProfiles: async (
    params?: ExportProfilesParams
  ): Promise<ExportResponse<CDPProfile> | Blob> => {
    const format = params?.format || 'json'

    if (format === 'csv') {
      const response = await apiClient.get('/cdp/export/profiles', {
        params: { ...params, format: 'csv' },
        responseType: 'blob',
      })
      return response.data
    }

    const response = await apiClient.get<ExportResponse<CDPProfile>>(
      '/cdp/export/profiles',
      { params }
    )
    return response.data
  },

  /**
   * Export events in JSON or CSV format
   * @param params Export parameters (format, limit, offset, date filters)
   */
  exportEvents: async (
    params?: ExportEventsParams
  ): Promise<ExportResponse<CDPEvent> | Blob> => {
    const format = params?.format || 'json'

    if (format === 'csv') {
      const response = await apiClient.get('/cdp/export/events', {
        params: { ...params, format: 'csv' },
        responseType: 'blob',
      })
      return response.data
    }

    const response = await apiClient.get<ExportResponse<CDPEvent>>(
      '/cdp/export/events',
      { params }
    )
    return response.data
  },
}

// =============================================================================
// React Query Hooks
// =============================================================================

// Query Keys
export const cdpQueryKeys = {
  all: ['cdp'] as const,
  profiles: () => [...cdpQueryKeys.all, 'profiles'] as const,
  profile: (id: string) => [...cdpQueryKeys.profiles(), id] as const,
  profileByIdentifier: (type: IdentifierType, value: string) =>
    [...cdpQueryKeys.profiles(), 'lookup', type, value] as const,
  sources: () => [...cdpQueryKeys.all, 'sources'] as const,
  health: () => [...cdpQueryKeys.all, 'health'] as const,
  webhooks: () => [...cdpQueryKeys.all, 'webhooks'] as const,
  webhook: (id: string) => [...cdpQueryKeys.webhooks(), id] as const,
  anomalies: () => [...cdpQueryKeys.all, 'anomalies'] as const,
  anomalySummary: () => [...cdpQueryKeys.anomalies(), 'summary'] as const,
}

/**
 * Hook to get a profile by ID
 */
export function useCDPProfile(profileId: string, options?: { enabled?: boolean }) {
  return useQuery({
    queryKey: cdpQueryKeys.profile(profileId),
    queryFn: () => cdpApi.getProfile(profileId),
    staleTime: 30 * 1000, // 30 seconds
    enabled: options?.enabled ?? !!profileId,
  })
}

/**
 * Hook to lookup a profile by identifier
 */
export function useCDPProfileLookup(
  identifierType: IdentifierType,
  identifierValue: string,
  options?: { enabled?: boolean }
) {
  return useQuery({
    queryKey: cdpQueryKeys.profileByIdentifier(identifierType, identifierValue),
    queryFn: () => cdpApi.lookupProfile(identifierType, identifierValue),
    staleTime: 30 * 1000,
    enabled: options?.enabled ?? (!!identifierType && !!identifierValue),
    retry: false, // Don't retry 404s
  })
}

/**
 * Hook to list all data sources
 */
export function useCDPSources() {
  return useQuery({
    queryKey: cdpQueryKeys.sources(),
    queryFn: () => cdpApi.listSources(),
    staleTime: 60 * 1000, // 1 minute
  })
}

/**
 * Hook to ingest events
 */
export function useIngestEvents(options?: { sourceKey?: string }) {
  const queryClient = useQueryClient()

  return useMutation({
    mutationFn: (events: EventInput[]) =>
      cdpApi.ingestEvents(events, options?.sourceKey),
    onSuccess: () => {
      // Invalidate profiles as they may have been updated
      queryClient.invalidateQueries({ queryKey: cdpQueryKeys.profiles() })
    },
  })
}

/**
 * Hook to ingest a single event
 */
export function useIngestEvent(options?: { sourceKey?: string }) {
  const queryClient = useQueryClient()

  return useMutation({
    mutationFn: (event: EventInput) =>
      cdpApi.ingestEvent(event, options?.sourceKey),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: cdpQueryKeys.profiles() })
    },
  })
}

/**
 * Hook to create a new data source
 */
export function useCreateSource() {
  const queryClient = useQueryClient()

  return useMutation({
    mutationFn: (source: SourceCreate) => cdpApi.createSource(source),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: cdpQueryKeys.sources() })
    },
  })
}

/**
 * Hook to check CDP health
 */
export function useCDPHealth() {
  return useQuery({
    queryKey: cdpQueryKeys.health(),
    queryFn: () => cdpApi.health(),
    staleTime: 60 * 1000,
    refetchInterval: 5 * 60 * 1000, // Every 5 minutes
  })
}

/**
 * Hook to list all webhooks
 */
export function useCDPWebhooks() {
  return useQuery({
    queryKey: cdpQueryKeys.webhooks(),
    queryFn: () => cdpApi.listWebhooks(),
    staleTime: 60 * 1000, // 1 minute
  })
}

/**
 * Hook to get a webhook by ID
 */
export function useCDPWebhook(webhookId: string, options?: { enabled?: boolean }) {
  return useQuery({
    queryKey: cdpQueryKeys.webhook(webhookId),
    queryFn: () => cdpApi.getWebhook(webhookId),
    staleTime: 30 * 1000,
    enabled: options?.enabled ?? !!webhookId,
  })
}

/**
 * Hook to create a new webhook
 */
export function useCreateWebhook() {
  const queryClient = useQueryClient()

  return useMutation({
    mutationFn: (webhook: WebhookCreate) => cdpApi.createWebhook(webhook),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: cdpQueryKeys.webhooks() })
    },
  })
}

/**
 * Hook to update a webhook
 */
export function useUpdateWebhook() {
  const queryClient = useQueryClient()

  return useMutation({
    mutationFn: ({ webhookId, update }: { webhookId: string; update: WebhookUpdate }) =>
      cdpApi.updateWebhook(webhookId, update),
    onSuccess: (_, { webhookId }) => {
      queryClient.invalidateQueries({ queryKey: cdpQueryKeys.webhooks() })
      queryClient.invalidateQueries({ queryKey: cdpQueryKeys.webhook(webhookId) })
    },
  })
}

/**
 * Hook to delete a webhook
 */
export function useDeleteWebhook() {
  const queryClient = useQueryClient()

  return useMutation({
    mutationFn: (webhookId: string) => cdpApi.deleteWebhook(webhookId),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: cdpQueryKeys.webhooks() })
    },
  })
}

/**
 * Hook to test a webhook
 */
export function useTestWebhook() {
  return useMutation({
    mutationFn: (webhookId: string) => cdpApi.testWebhook(webhookId),
  })
}

/**
 * Hook to rotate webhook secret
 */
export function useRotateWebhookSecret() {
  const queryClient = useQueryClient()

  return useMutation({
    mutationFn: (webhookId: string) => cdpApi.rotateWebhookSecret(webhookId),
    onSuccess: (_, webhookId) => {
      queryClient.invalidateQueries({ queryKey: cdpQueryKeys.webhook(webhookId) })
    },
  })
}

/**
 * Hook to detect event volume anomalies
 */
export function useEventAnomalies(params?: AnomalyDetectionParams) {
  return useQuery({
    queryKey: [...cdpQueryKeys.anomalies(), params],
    queryFn: () => cdpApi.detectEventAnomalies(params),
    staleTime: 5 * 60 * 1000, // 5 minutes
  })
}

/**
 * Hook to get anomaly detection summary
 */
export function useAnomalySummary() {
  return useQuery({
    queryKey: cdpQueryKeys.anomalySummary(),
    queryFn: () => cdpApi.getAnomalySummary(),
    staleTime: 60 * 1000, // 1 minute
    refetchInterval: 5 * 60 * 1000, // Every 5 minutes
  })
}

// =============================================================================
// Utility Functions
// =============================================================================

/**
 * Create a track event helper for easy event ingestion
 */
export function createTracker(sourceKey?: string) {
  return {
    track: async (
      eventName: string,
      identifiers: IdentifierInput[],
      properties?: Record<string, unknown>,
      options?: {
        idempotencyKey?: string
        context?: EventContext
        consent?: EventConsent
      }
    ): Promise<EventIngestResult> => {
      const event: EventInput = {
        event_name: eventName,
        event_time: new Date().toISOString(),
        idempotency_key: options?.idempotencyKey,
        identifiers,
        properties,
        context: options?.context,
        consent: options?.consent,
      }
      return cdpApi.ingestEvent(event, sourceKey)
    },

    identify: async (
      identifiers: IdentifierInput[],
      traits?: Record<string, unknown>
    ): Promise<EventIngestResult> => {
      const event: EventInput = {
        event_name: 'Identify',
        event_time: new Date().toISOString(),
        identifiers,
        properties: traits ? { traits } : undefined,
      }
      return cdpApi.ingestEvent(event, sourceKey)
    },

    page: async (
      identifiers: IdentifierInput[],
      pageName?: string,
      properties?: Record<string, unknown>,
      context?: EventContext
    ): Promise<EventIngestResult> => {
      const event: EventInput = {
        event_name: 'PageView',
        event_time: new Date().toISOString(),
        identifiers,
        properties: { page_name: pageName, ...properties },
        context,
      }
      return cdpApi.ingestEvent(event, sourceKey)
    },
  }
}
